### PR TITLE
Add SANA-WM integration (_bidirectional and _streaming variants)

### PR DESCRIPTION
**Pull request by aidanfnv**
## Summary

Adds the `flashdreams-sana-wm` integration for both public NVlabs/Sana camera-controlled world-model releases:

| upstream release | FlashDreams runner | inference contract | benchmark scope |
| --- | --- | --- | --- |
| `SANA-WM_bidirectional` | `sana-wm-bidirectional` | full-sequence I2V clip generation in one bidirectional pass | BF16 upstream-vs-FlashDreams clip latency |
| `SANA-WM_streaming` | `sana-wm-streaming` | chunk-causal autoregressive generation with streaming refiner/VAE decode | BF16, FP8, and FP4 upstream-vs-FlashDreams chunk latency |

The two upstream releases have distinct runner slugs, configs, checkpoint roots, parity helpers, benchmark scripts, benchmark data, and model cards.

## Implementation

- `SANA-WM_bidirectional` is wired through a native FlashDreams Stage-1 DiT, the LTX-2 refiner, VAE decode, camera conditioning, output writing, and the shared FlashDreams runner/pipeline/diffusion contracts.
- `SANA-WM_streaming` is wired as its own chunk-causal path with streaming Stage-1 conditioning, streaming checkpoint roots, chunk boundaries, streaming LTX-2 refiner settings, causal VAE decode, and progressive output emission.
- Runner configs select separate bidirectional and streaming checkpoint roots for Stage-1, refiner, VAE, and text-encoder assets.
- Runtime quantization does not depend on TransformerEngine. FP8/FP4 paths use Torch `_scaled_mm` plus integration-local Triton helpers where needed; no upstream fused kernels are vendored.
- `SANA-WM_bidirectional` exposes FlashDreams FP8/FP4 options, but the matched upstream comparison is BF16-only because upstream `SANA-WM_bidirectional` does not expose the streaming release's FP8/FP4 precision flags.
- `SANA-WM_streaming` owns the matched BF16/FP8/FP4 upstream comparisons.

## User-Facing Changes

- New integration package under `integrations/sana/`, published as `flashdreams-sana-wm` (Python package `sana_wm`).
- Two registered runner slugs:
  - `sana-wm-bidirectional`
  - `sana-wm-streaming`
- Separate docs pages:
  - `docs/source/models/sana_wm_bidirectional.rst`
  - `docs/source/models/sana_wm_streaming.rst`
  - Both are linked from the landing page and model index, each with a sample AVIF clip.
- Separate chart data:
  - `docs/source/_static/performance/sana_wm_bidirectional/perf-0801-bf16.md`
  - `docs/source/_static/performance/sana_wm_streaming/perf-0801-bf16.md`
  - `docs/source/_static/performance/sana_wm_streaming/perf-0801-fp8.md`
  - `docs/source/_static/performance/sana_wm_streaming/perf-0801-fp4.md`
- Camera inputs can come from a pose file or the action DSL, for example `"w-100,dw-60,w-101"`.
- `--intrinsics-path` is optional. When omitted, intrinsics are derived from the first-frame geometry and `--intrinsics-hfov-deg`.
- Explicit camera paths and generated action trajectories are fitted to the requested frame count. Streaming requests are snapped to `8 * --num-frame-per-block * k + 1`.

## Implementation Notes

- The Stage-1 DiT implementation lives in `integrations/sana/sana_wm/` and covers both bidirectional and chunk-causal streaming transformer variants.
- Bidirectional and streaming use distinct constants for Hugging Face repos, config roots, Stage-1 checkpoints, refiner roots, and VAE roots.
- `_stage1_state_dict()` normalizes the two release formats into one state dict: bidirectional ships flat safetensors, streaming ships a `.pt` nesting the tensors under `generator` / `state_dict` with `model.`-prefixed keys. Both load under `strict=True`.
- Streaming uses a FlashDreams-owned config literal instead of vendoring the upstream YAML.
- `DiffusionModel.run()` delegates initial latent creation to `Transformer.initial_noise()`, so SANA-WM first-frame pinning can live in the transformer boundary without a SANA-only diffusion-model subclass.
- `integrations/sana/tests/parity_check/` adds a correctness harness that dispatches through `SANA_WM_VARIANT=bidirectional|streaming`, with separate FlashDreams helpers for each release.
- `integrations/sana/tests/performance/` holds the timing harness (`bench.sh`, `bench_summary.py`, `bench_sweep_summary.py`). It clones and pins the upstream Sana checkout, applies instrumentation-only patches, and runs both sides from an isolated `uv` environment. The benchmark summary code rejects low-precision upstream comparisons for `SANA-WM_bidirectional` and treats streaming metrics as steady-state milliseconds per produced chunk.
- TransformerEngine appears only in the benchmark harness's optional `quant` extra, used to run the *upstream streaming* FP8/FP4 side. FlashDreams low-precision rows use the base harness environment, and the FlashDreams runtime path never imports it.

## Differences From Official SANA-WM Code

- **Low-precision backend**: official SANA-WM uses TransformerEngine for FP8/FP4 paths (`te.Linear`, `te.fp8_autocast`, `NVFP4BlockScaling`, and `Float8BlockScaling`). The FlashDreams integration uses Torch/Triton replacement layers instead.
- **Intrinsics fallback**: omitted intrinsics are derived from a fixed-HFoV assumption instead of Pi3X. This is lighter-weight, but less accurate when the true camera HFoV differs from the assumption.
- **No prepared-module cache**: upstream caches the pickled post-quantization Stage-1 module and reloads it on subsequent FP4 runs. The integration re-quantizes on every launch. This affects FP4 startup only — BF16 and FP8 do not use the cache upstream, and preparation happens outside the benchmark timing boundary either way.

## Benchmarks

Published on the model cards. One NVIDIA GB300 GPU, upstream Sana at `6298508`, FlashDreams at `bd0816e` (later commits have been smoke tested to ensure they still produce the same results), one discarded warmup run and three measured runs. Model construction, checkpoint loading, video writing, and frame dumps are outside the timing boundary.

`SANA-WM_bidirectional`, BF16, 121-frame full-pipeline clip (Stage-1 DiT + LTX-2 refiner + VAE decode):

| metric | official | FlashDreams |
| --- | ---: | ---: |
| post-load generation latency / clip | 56,932.83 ms | 34,182.39 ms |

`SANA-WM_streaming`, steady-state generation latency per produced chunk (241 requested frames, `w-80,dw-40,w-80,aw-40` action path):

| precision | official | FlashDreams |
| --- | ---: | ---: |
| BF16 | 1,170.29 ms/chunk | 1,957.93 ms/chunk |
| FP8 | 1,482.92 ms/chunk | 2,392.67 ms/chunk |
| FP4 | 1,594.33 ms/chunk | 4,118.36 ms/chunk |

The streaming rows are published as a measured baseline rather than a claimed speedup: the FlashDreams streaming path is functional and benchmarked, but still slower than the official implementation.

## Test Plan

Setup:

```bash
uv sync --package flashdreams-sana-wm --extra dev
```

CPU tests (113 total):

```bash
uv run --package flashdreams-sana-wm --extra dev pytest \
    integrations/sana/tests/test_smoke.py \
    integrations/sana/tests/test_camera.py \
    integrations/sana/tests/test_performance_benchmark_summary.py \
    integrations/sana/tests/test_parity_quality_tools.py
```

| suite | tests | covers |
| --- | ---: | --- |
| `test_smoke.py` | 76 | runner/pipeline/config wiring, both transformer variants, conditioning, decode, refiner adapter, scheduler, quant layer construction, HF path resolution, and the pinned artefact defaults |
| `test_camera.py` | 14 | pose-file and action-DSL trajectories, bounded action materialization, frame-count fitting, and intrinsics derivation |
| `test_performance_benchmark_summary.py` | 20 | benchmark summary parsing, precision-row rejection rules, SANA benchmark `quant` extra routing, and streaming steady-state metrics |
| `test_parity_quality_tools.py` | 3 | parity diff and streaming-continuity helpers |

GPU tests (27 total, run on RTX PRO 6000 Blackwell):

```bash
uv run --package flashdreams-sana-wm --extra dev pytest \
    integrations/sana/tests/test_stage1_cuda.py \
    integrations/sana/tests/test_quant_cuda.py
```

| suite | tests | covers |
| --- | ---: | --- |
| `test_stage1_cuda.py` | 24 | Stage-1 CUDA execution, chunk-causal streaming windows, and KV-cache behaviour |
| `test_quant_cuda.py` | 3 | FP8/FP4 `_scaled_mm` and Triton quantization kernels |

End-to-end generation, both runners, BF16, `demo_0` assets:

- `sana-wm-bidirectional` writes a clip; Stage-1 weights load under `strict=True` with no missing or unexpected keys.
- `sana-wm-streaming` writes a clip; same clean `strict=True` load of the `.pt` release.

Full benchmark sweeps (BF16/FP8/FP4, both variants, against the pinned upstream Sana checkout at `6298508`) were run on RTX PRO 6000 Blackwell and on GB300. The model cards publish the GB300 rows.

### CODE REVIEW

File: integrations/sana/sana_wm/camera.py [hunk 1]
```diff
@@ -0,0 +1,610 @@
...
+            keys: list[str] = []
+        else:
+            bad = sorted({key for key in keys_lower if key not in ALLOWED_ACTION_KEYS})
+            if bad:
+                raise ValueError(
+                    f"Action segment {segment!r} contains unknown keys {bad}; "
+                    f"allowed: {''.join(sorted(ALLOWED_ACTION_KEYS))}."
+                )
+            keys = sorted(set(keys_lower))
+        per_frame.extend([keys] * int(duration_part))
```
[the remark this answers is not in the document]
Comment (aidanfnv) [reply integrations/sana/sana_wm/camera.py]: Fixed

**Review by jarcherNV — APPROVED** [review]
This was pretty long so admittedly I only skimmed it, but I didn't notice any glaring issues so I think it is fine to merge.

**Comment by jarcherNV** [comment]
An additional comment: There are a lot of commits in this PR, so it may produce a pretty long commit message when merged. You may want to consider squashing it and creating a reasonable commit message for everything, but I'll leave it up to you to decide how to proceed.
